src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,291 @@
+                                }
+
+                                if (!File.Exists(projectRefInfo.OutputFilePath))
+                                {
+                                    metadataReferenceSet.Remove(projectRefInfo.OutputFilePath);
+                                }
+                            }
+                            else
+                            {
+                                projectReferences.Add(CreateProjectReference(id, projectRefInfo.Id, projectFileReference.Aliases));
+                                metadataReferenceSet.Remove(projectRefInfo.OutputRefFilePath);
+                                metadataReferenceSet.Remove(projectRefInfo.OutputFilePath);
+                            }
+
+                            continue;
+                        }
+                    }
+
+                    // We weren't able to handle this project reference, so add it without further processing.
+                    var unknownProjectId = _projectMap.GetOrCreateProjectId(projectFileReference.Path);
+                    projectReferences.Add(CreateProjectReference(id, unknownProjectId, projectFileReference.Aliases));
+                }
+
+                return new ResolvedReferences(
+                    projectReferences.ToImmutableArray(),
+                    metadataReferenceSet.GetRemaining());
+            }
+
+            private ProjectReference CreateProjectReference(ProjectId from, ProjectId to, ImmutableArray<string> aliases)
+            {
+                var result = new ProjectReference(to, aliases);
+                if (!_projectIdToProjectReferencesMap.TryGetValue(from, out var references))
+                {
+                    references = new List<ProjectReference>();
+                    _projectIdToProjectReferencesMap.Add(from, references);
+                }
+
+                references.Add(result);
+
+                return result;
+            }
+
+            private bool ProjectReferenceExists(ProjectId to, ProjectId from)
+                => _projectIdToProjectReferencesMap.TryGetValue(from, out var references)
+                && references.Contains(pr => pr.ProjectId == to);
+
+            /// <summary>
+            /// Try to get the <see cref="ProjectId"/> for an already loaded project whose output path is in the given <paramref name="metadataReferenceSet"/>.
+            /// </summary>
+            private bool TryAddProjectReferenceToLoadedOrMappedProject(ProjectId id, string projectReferencePath, MetadataReferenceSet metadataReferenceSet, ImmutableArray<string> aliases, List<ProjectReference> projectReferences)
+            {
+                if (_projectMap.TryGetIdsByProjectPath(projectReferencePath, out var mappedIds))
+                {
+                    foreach (var mappedId in mappedIds)
+                    {
+                        // Don't add a reference if the project already has a reference on us. Otherwise, it will cause a circularity.
```

Comment:
This isn't checking for transitive circularities?

Reply:
Nope. MSBuildWorkspace never did, as far as I can tell.